ledger/simulation/resources.go

```diff
@@ -605,3 +607,698 @@ func (p *resourcePolicy) AvailableBox(app basics.AppIndex, name string, operatio
+			err := p.txnResources[i].addAsset(asset)
+			if err != nil {
+				return err
+			}
+		}
+
+		for _, app := range sortedApps {
+			err := p.txnResources[i].addApp(app)
+			if err != nil {
+				return err
+			}
+		}
+
+		for _, account := range sortedAccounts {
+			err := p.txnResources[i].addAccount(account)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Then assign cross-reference resources because they have the most strict requirements (one account and another resource)
+	for _, holding := range groupResources.AssetHoldings {
+		err := p.addHolding(holding.Address, holding.Asset)
+		if err != nil {
+			return err
+		}
+
+		// Remove the resources from the global tracker in case they were added separately
+		groupResources.Assets = slices.DeleteFunc(groupResources.Assets, func(a basics.AssetIndex) bool {
+			return a == holding.Asset
+		})
+		groupResources.Accounts = slices.DeleteFunc(groupResources.Accounts, func(a basics.Address) bool {
+			return a == holding.Address
+		})
+	}
+
+	for _, local := range groupResources.AppLocals {
+		err := p.addLocal(local.Address, local.App)
+		if err != nil {
+			return err
+		}
+
+		// Remove the resources from the global tracker in case they were added separately
+		groupResources.Apps = slices.DeleteFunc(groupResources.Apps, func(a basics.AppIndex) bool {
+			return a == local.App
+		})
+		groupResources.Accounts = slices.DeleteFunc(groupResources.Accounts, func(a basics.Address) bool {
+			return a == local.Address
+		})
+	}
+
+	// Then assign boxes because they can take up to two slots
+	for _, box := range groupResources.Boxes {
+		err := p.addBox(box.App, box.Name)
+		if err != nil {
+			return err
+		}
+
+		// Remove the app from the global tracker in case it was added separately
+		groupResources.Apps = slices.DeleteFunc(groupResources.Apps, func(a basics.AppIndex) bool {
+			return a == box.App
+		})
+	}
+
+	// Then assign accounts because they have a lower limit than other resources
+	for _, account := range groupResources.Accounts {
+		err := p.addAccount(account)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Finally assign the remaining resources which just require one slot
+	for _, app := range groupResources.Apps {
+		err := p.addApp(app)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, asset := range groupResources.Assets {
+		err := p.addAsset(asset)
+		if err != nil {
+			return err
+		}
+	}
+
+	for i := 0; i < groupResourceTracker.NumEmptyBoxRefs; i++ {
+		err := p.addBox(0, "")
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *resourcePopulator) getPopulatedArrays() map[int]PopulatedResourceArrays {
+	populatedArrays := map[int]PopulatedResourceArrays{}
+	for _, i := range p.appCallIndexes {
+		resources := p.txnResources[i]
+		if resources == nil {
+			continue
+		}
+
+		pop := resources.getPopulatedArrays()
+		if i >= p.groupSize && len(pop.Accounts)+len(pop.Assets)+len(pop.Apps)+len(pop.Boxes) == 0 {
+			break
+		}
```

Comment:
could you explain this condition? It appears checks existence of extra resources?

Reply:
Added comment in [413893f](https://github.com/algorand/go-algorand/pull/6015/commits/413893f7ad8ff6140d332109ee7859edb6d9ddde). Once all of the group's txns has been populated, once we hit all empty resources we know we can stop since there won't be any more